### ISSUE TITLE
Optimize macro-synteny services for fragmented assemblies

### ISSUE THREAD

**Issue by alancleary** [legumeinfo/microservices#48]
@adf-ncgr has reported that the macro-synteny services can have very long running times when computing synteny blocks on databases containing many fragmented assemblies (scaffolds, contigs, super contigs). The goal of this issue is to fix this problem, though the solution is currently unclear.

So far I've created a `macro-synteny-performance` branch and added a chromosome length filter to the macro-synteny service. This has virtually the same performance on chromosome level assemblies and, as reported by @adf-ncgr, has improved performance on fragmented assemblies. However, the improved performance is still lacking. Next I'll profile the code and see if there's any other apparently optimizations to be made.

**Comment by adf-ncgr** [comment]
to be clear (for my sake) I think when you say here "added a chromosome length filter" I think you are talking about the gene count filter? chromosome lengths being a different attribute that I was filtering on earlier and which will likely also be a useful option for users to control, but currently I think we aren't talking so much about user-specifiable filters (which will come into play with legumeinfo/gcv#401) as about things we can do to eliminate unneeded overhead without the prospect of missing possible matches for the sake of performance. Right?

**Comment by alancleary** [comment]
After profiling both the macro-synteny and pairwise macro-synteny services it's apparent that the significant bottlenecks that remain are related to getting data from Redis, i.e. most of the execution time spent is spent waiting for responses from Redis. I suspect this is because Redis is spending lots of time encoding the chromosomes (large ordered lists of genes). I'll look into whether there's any optimizations that would actually be reasonable to implement.

**Comment by adf-ncgr** [comment]
@alancleary this result puzzles me a little; how do the redis access patterns in this case differ from what is done to find microsynteny (which seems to be relatively fast)? Naively, I would have thought you'd need to get the same data out of redis in both cases. I should be mostly free this afternoon if you feel like discussing in real time.

**Comment by alancleary** [comment]
> how do the redis access patterns in this case differ from what is done to find microsynteny (which seems to be relatively fast)?

The query sent to the macro-synteny services is potentially thousands of genes long, whereas the query sent to the micro-synteny service is at most a few hunderd genes, so the services use different strategies for pulling data from the database. The micro-synteny service asks for all genes that have the same families as the query and builds those genes into blocks based on proximity; a bottom-up approach. Conversely, the macro-synteny services just pulls in all genes for each chromosome and then derives blocks from subsequences; a top-down approach.

I guess if we don't find a reasonable Redis optimizations then it might be worth prototyping a version of the macro-synteny services that uses a similar approach to the micro-synteny service, but this would consume a bit a time so we'd probably want to consider other priorities before diving in.

**Comment by alancleary** [comment]
> I guess if we don't find a reasonable Redis optimizations then it might be worth prototyping a version of the macro-synteny services that uses a similar approach to the micro-synteny service, but this would consume a bit a time so we'd probably want to consider other priorities before diving in.

Now that I've said this, a middle ground solution may be to use this strategy to filter what chromosomes we attempt to compute macro-synteny on, i.e. pull in the genes that have families in the query chromosome, group them by chromosome, and then filter these gene groups using the matched parameter, leaving only chromosomes that are much more likely to have blocks.

**Comment by adf-ncgr** [comment]
Thanks, the top down vs bottom up difference makes sense. Regarding the middle ground, I guess whether that would make sense might depend on how well querying by family scales with the number of families to be queried. But, I agree we probably ought to be considering other priorities if there's no relatively easy fix for our worst case scenario databases.

**Comment by alancleary** [comment]
OK. I just went ahead and prototyped the middle ground solution I outlined. Here are the service stats before:

targets identified: 54346  (all chromosomes in the database)
filtered targets: 40047  (chromosomes with length > match param)
targets with blocks: 10
run-time: 53143ms

And after:

targets identified: 16147  (all chromosomes with at least one gene matching a query family)
filtered targets: 346  (all chromosomes with more genes matched than the matched param)
targets with blocks: 10
run-time: 4153

I think the filtered targets could be further reduced be considering the matched gene positions and the max insertion param, although at this point we'd be somewhat on our way to the full-blown reimplementation I originally mentioned...

**Comment by adf-ncgr** [comment]
That's a very nice result, assuming the query chromosome here was a "normal" sized one. 4 seconds seems perfectly acceptable to me to get macrosynteny returned, but I'll check out macro-synteny-performance and see how it behaves on some other dbs before we declare your victory. Thanks, and have a great weekend!

**Comment by alancleary** [comment]
Actually, it was a scaffold. I guess a full chromosome is a better measure of cost-benefit; here's the same stats for a full chromosome.

Before:

targets identified: 54346
filtered targets: 40047
targets with blocks: 452
run-time: 255639ms

After:

targets identified: 24061
filtered targets: 3155
targets with blocks: 452
run-time: 112260ms

It's still a large improvement but the run-time isn't as practical. Clearly the bottleneck is a function of how many chromosomes are actually being copied (confirmed with profiling), so I think taking the intermediate parameter into consideration is also worthwhile. If we go that far, though, we might as well just refactor the service to compute on these "sparse chromosomes" rather than pulling in the entire chromosome to recompute a lot of what we already know.

**Comment by adf-ncgr** [comment]
I think that makes sense- perhaps you implied this already but it almost seems like the macrosynteny-service would effectively be working like the microsynteny-service, at least in terms of defining regions on the targets is concerned (just that the query would not be limited by neighborhood)?

**Comment by alancleary** [comment]
Yep. Perhaps the services could be combined...

**Comment by adf-ncgr** [comment]
my sentiments exactly.  And not strictly related, but it might be worth also thinking about implications for the multi-alignment service, in the sense that we have an open issue somewhere about edge effects caused by the fact that using the same neighborhood for each segment tends to exclude some genes that really ought to be included (due to insertions in some segments making their ideal neighborhoods "longer"). Seems like collecting all the gene families from the minimal neighborhoods (ie whatever is specified by the user) across all starting segments and then finding the maximal neighborhood for each segment that includes matches to any of the combined gene families from the first pass might be the way to go for that.

**Comment by adf-ncgr** [comment]
maybe the more relevant open issue is legumeinfo/gcv#77 which as you noted in legumeinfo/gcv#269 has been a long time coming...

**Comment by alancleary** [comment]
I agree, this all seems to be related. For the sake of this issue's scope and the site currently afflicted by this issue, I'll implement the solution I prototyped here and we can pursue the broader issues in an epic.

**Comment by alancleary** [comment]
OK. I updated the target chromosome filtration to use both the "match" and "intermediate" parameters. The intermediate parameter further reduces the overall run-time: 300005ms unfiltered, 231986ms match filtered, and 78820ms match+intermediate filtered. I need to add support for filtering pre-defined lists of targets, which will have to wait until tomorrow. Then I'll push up the code.

It's worth noting that the bottleneck is now how the data are organized in the database. I think the run-times could be greatly improved by expanding the schema, but this will take quite a bit of test-driven development.
